swmmio/tests/test_model_elements.py

```diff
@@ -171,4 +171,45 @@ def test_get_set_timeseries(test_model_02):
     assert(ts.loc['TS2'].Date == 'FILE')
     assert('"' in ts.loc['TS2'].Value)
     assert(ts.Value.isnull().sum() == 0)
-    print (ts)
\ No newline at end of file
+    print (ts)
+
+
+def test_inp_tags_df():
+    expected_output = pd.DataFrame({'ElementType': ['Subcatch'] * 4,
+                                    'Name': ['CA-1', 'CA-7', 'CA-8', 'CA-11'],
+                                    'Tag': ['CA'] * 4})
+    expected_output.set_index(['ElementType'], inplace=True)
+
+    model = Model(MODEL_GREEN_AMPT)
+    actual_output = model.inp.tags
+
+    assert expected_output.equals(actual_output.sort_index())
+
+
+def test_inp_tags_getter_and_setter():
+    tags_to_set = pd.DataFrame({'ElementType': ['Subcatch'] * 4,
+                                'Name': ['CA-1', 'CA-7', 'CA-8', 'CA-11'],
+                                'Tag': ['CA'] * 4})
+    tags_to_set.set_index(['ElementType'], inplace=True)
+
+    model = Model(MODEL_GREEN_AMPT)
+
+    with tempfile.TemporaryDirectory() as tempdir:
+        temp_inp_path = os.path.join(tempdir, f'{model.inp.name}.inp')
+        model.inp.save(temp_inp_path)
+        temp_model = Model(temp_inp_path)
+        temp_model.inp.tags = tags_to_set
+        retrieved_tags = temp_model.inp.tags
+
+    assert tags_to_set.equals(retrieved_tags.sort_index())
+
+
+def test_model_tags():
+    expected_output = pd.DataFrame({'ElementType': ['Subcatch'] * 4,
+                                    'Name': ['CA-1', 'CA-7', 'CA-8', 'CA-11'],
+                                    'Tag': ['CA'] * 4})
+    expected_output.set_index(['ElementType'], inplace=True)
+
+    model = Model(MODEL_GREEN_AMPT)
+    actual_output = model.tags
```

Comment:
@BuczynskiRafal I think the reason this fails is because we don't actually have a `tags` property in the `swmmio.Model` object. 

The `Model` object is a higher level abstraction whose properties bring together multiple related tables in the inp file. For example, the `Model.nodes` property combines the [JUNCTIONS], [OUTFALLS], and [STORAGE] tables. However, I don't think this higher-level abstraction is necessary in the case of tags. Is there a use-case for this that I'm not thinking of?

That said, instead of adding a `tags` property to the `swmmio.Model` object, we could certainly include tags _within_ the `Model.nodes` section. This way, we when we want fetch all the relevant information about nodes in our models, we can include any tags that may be assigned in the dataframe that is returned by `Model.nodes.dataframe`. 

Does that make sense to you?

Reply:
Yes, I agree with you. I left the setter and getter for the inp class. 

I'm don't know how to solve tag assignment using the approach you propose i.e. by referencing the key from the "section_headers.yml" file. 
The tag section is indexed by the element type and not by its name/id. 

```txt
[TAGS]
Subcatch  CA-1             CA              
Subcatch  CA-7             CA              
Subcatch  CA-8             CA              
Subcatch  CA-11            CA
Link      P001             CA
```
I modified the yaml
```yaml
  nodes:
    inp_sections: [junctions, outfalls, storage, tags]
    rpt_sections: [Node Depth Summary, Node Flooding Summary, Node Inflow Summary]
    columns: [InvertElev, MaxDepth, InitDepth, SurchargeDepth, PondedArea,
               OutfallType, StageOrTimeseries, TideGate, MaxD, StorageCurve,
               Coefficient, Exponent, Constant, EvapFrac, SuctionHead,
               Conductivity, InitialDeficit, coords, tag]
```
output
```txt
          InvertElev  MaxDepth  ...  InitialDeficit                coords
Link             NaN       NaN  ...             NaN                  None
Outlet          0.00       NaN  ...             NaN      [(500.0, 500.0)]
P001            1.10      20.0  ...             NaN      [(100.0, 100.0)]
P005            1.58       NaN  ...            0.12      [(200.0, 200.0)]
P009            3.60      20.0  ...             NaN  [(498.395, 215.716)]
P011            2.89      20.0  ...             NaN      [(400.0, 400.0)]
Subcatch         NaN       NaN  ...             NaN                  None
Subcatch         NaN       NaN  ...             NaN                  None
Subcatch         NaN       NaN  ...             NaN                  None
Subcatch         NaN       NaN  ...             NaN                  None
```

This worked with conduits method 
```python
        tags = dataframe_from_inp(inp.path, "TAGS")
        if "Link" in set(tags.index):
            df = df.merge(dataframe_from_inp(inp.path, "Tags"), left_on="Name", right_on="Name", how="left")
```
output
```text
   Name InletNode  ... SlopeFtPerFt  Tag
0  P001      P001  ...        0.005   CA
1  P005      P005  ...        0.002  NaN
2  P009      P009  ...        0.004  NaN
3  P011      P011  ...        0.002  NaN
```